### ISSUE TITLE
scalabilityMode: Same as in getParameters() or not?

### ISSUE THREAD

**Issue by henbos** [w3c/webrtc-stats#731]
It was brought to my attention that if you ask for "L3T3_KEY" the encoder may still drop layers for adaptation reasons.
If I understand correctly, this means that getParameters() could say "L3T3_KEY" (just like you had previously set it), but getStats() might say "L2T3_KEY" or even "L1T3_KEY".

This surprised me, but at the same time, I think it makes sense that getStats() says what is currently happening, whether or not that is a subset of what was set in setParameters().

Qs:
- Do we agree this makes sense?
- Does [the spec](https://w3c.github.io/webrtc-stats/#dom-rtcoutboundrtpstreamstats-scalabilitymode) need to be clarified or should we just close this issue?

**Comment by Orphis** [comment]
While I don't think it necessarily makes sense in general, this is the best information we currently have about the current state of SVC in the only shipping implementation.

Since we could not agree on SVC-layer stats at TPAC in 2019 (issues regarding the layer id) it might be acceptable in the meantime. But we should really have interested parties work on SVC related stats to sort this out properly.

**Comment by alvestrand** [comment]
I don't think this is reasonable - there is no guarantee that the result of dropping layers out of a named configuration is a named configuration. If it isn't, what could one possibly say?

I think we need to get stats for SVC (currently active layers) spec'ed, not hack on the getStats() interface.

**Comment by eshrubs** [comment]
In my opinion, the scalability mode is that which the encoder is configured with. If the frame is dropped due to some rate limit then it does not affect the encoding structure, and if it is dropped in egress then it is also not affecting the encoding structure but rather a dropped frame in the network. 

So with the point-of-view of the encoder, if you reconfigure to drop layers, you are going from S spatial layers to S-1 spatial layers, then it will have a defined name (L{S}Tx => L{S-1}Tx). Same thing with temporal layers, T -> T-1 temporal layers has a defined name (LxT{T} -> LxT{T-1}). You can never drop the last layer, and there can't be more than 3 layers, so they are always named.

Technically a codec implementation could produce a structure which is not named, for exampling if scaling was neither 2 nor 1.5, or with some very custom spatial or temporal scaling. This is not a real problem in practice since there is no way to configure that to happen.

**Comment by alvestrand** [comment]
Trying to find complex cases: There is an L3T2_KEY_SHIFT, but no L3T1_KEY_SHIFT. So if we deactivate a temporal layer in L3T2_KEY_SHIFT, do we end up in L3T2_KEY? And if we reactivate it, do we end up in L3T2_KEY_SHIFT?

It seems simpler to return the configured scalability mode.

**Comment by henbos** [comment]
Care to weigh in your thoughts on this issue, @ssilkin ?

Expose same value as in getParameters to avoid risk of no matching enum or expose a subset of the param one where applicable?

**Comment by eshrubs** [comment]
I am unsure as to why L3T1_KEY is missing from the table? It is present in the diagrams https://www.w3.org/TR/webrtc-svc/#L3T1_KEY*

An algorithm for decrementing a spatial layer would be,

```
decrementSpatialLayer(svc)
  if (svc == LxTy) return L(x-1)Ty
  if (svc == SxTy) return S(x-1)Ty if (x-1 != 1) else L1Ty
  if (svc == LxTy_KEY or LxTy_KEY_SHIFT) return L(x-1)Ty_KEY(_SHIFT) if (x-1 != 1) else L1Ty
```

and for temporal layers

```
decrementTemporalLayer(svc)
  if (svc == LxTy) return LxT(y-1)
  if (svc == SxTy) return SxT(y-1)
  if (svc == LxTy_KEY) return LxT(y-1)_KEY
  if (svc == LxTy_KEY_SHIFT) return LxT(y-1)_KEY_SHIFT if (y-1 != 1) else LxT1_KEY
```

**Comment by henbos** [comment]
Seems like a mistake indeed, I filed https://github.com/w3c/webrtc-svc/issues/81 for that missing scalabilityMode

**Comment by aboba** [comment]
@henbos @eshrubs AV1 bitstream spec does not mention `L3T1_KEY`.  So while the diagram makes sense (and we could certainly add the node to the table), I'm wondering if `L3T1_KEY` is supported in Chromium.  See: https://github.com/w3c/webrtc-svc/pull/82

**Comment by alvestrand** [comment]
> I am unsure as to why L3T1_KEY is missing from the table? It is present in the diagrams https://www.w3.org/TR/webrtc-svc/#L3T1_KEY*
> 
> An algorithm for decrementing a spatial layer would be,
> 
> ```
> decrementSpatialLayer(svc)
>   if (svc == LxTy) return L(x-1)Ty
>   if (svc == SxTy) return S(x-1)Ty if (x-1 != 1) else L1Ty
>   if (svc == LxTy_KEY or LxTy_KEY_SHIFT) return L(x-1)Ty_KEY(_SHIFT) if (x-1 != 1) else L1Ty
> ```

And what's the algorithm for incrementing? That one is the one that gets confusing because of the _SHIFT.

**Comment by henbos** [comment]
Now that https://github.com/w3c/webrtc-svc/issues/81 was fixed it sounds like decrementing should work. For incrementing, I guess the encoder would know how to reverse the degradations it performed?

**Comment by henbos** [comment]
Regarding

> I'm wondering if L3T1_KEY is supported in Chromium

@DanilChapovalov responded in https://github.com/w3c/webrtc-svc/issues/81
> L3T1_KEY is a valid scalability mode. Chrome should already support it both for AV1 and VP9.

**Comment by henbos** [comment]
@Orphis can you confirm my latest comment that getParameters() reflect what scalabilityMode we will attempt, not what scalabilityMode was the result in the event of a fallback?

**Comment by henbos** [comment]
Hmm, no fallback does get surfaced (see [tests](https://webrtc-review.googlesource.com/c/src/+/295872)), but there are cases where getParameters()'s scalabilityMode is undefined but getStats() returns L1T1.

**Comment by Orphis** [comment]
`getParameters()` should only return the current `scalabilityMode` when one was explicitly requested in the first place. So it should only change during a fallback.

**Comment by henbos** [comment]
I think [this](https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/modules/video_coding/codecs/vp9/svc_config.cc;l=187;drc=8101d6d81854e4f437e0f03f77694a01aaa8df7a) is why we sometimes see different scalabilityModes in getStats() than what we specified. This is not the regular fallback path but a special cap that happens when the input resolution is too low.

**Comment by henbos** [comment]
I think this reflecting the currently configured scalability mode is still accurate, maybe we can close this issue
